### PR TITLE
Improve TLS1.3 detection in registry for QUIC

### LINKED ISSUE #70673: Large numbers of Quic tests failed when run locally
I ran all inner+outerloop tests on my Windows 11 laptop (21H2 22000.675). 

770 System.Net.Quic tests failed with "SEC_E_ALGORITHM_MISMATCH | The client and server cannot communicate, because they do not possess a common algorithm."
```
    <test name="System.Net.Quic.Tests.QuicConnectionTests_MsQuicProvider.Dispose_WithPendingAcceptAndConnect_PendingAndSubsequentThrowOperationAbortedException" type="System.Net.Quic.Tests.QuicConnectionTests_MsQuicProvider" method="Dispose_WithPendingAcceptAndConnect_PendingAndSubsequentThrowOperationAbortedException" time="0.0306188" result="Fail">
        <failure exception-type="Microsoft.Quic.MsQuicException">
          <message><![CDATA[Microsoft.Quic.MsQuicException : ConfigurationLoadCredential failed: Unknown status '-2146893007']]></message>
          <stack-trace><![CDATA[   at Microsoft.Quic.MsQuic.ThrowIfFailure(Int32 status, String message) in C:\git\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\Interop\msquic.cs:line 47
   at System.Net.Quic.Implementations.MsQuic.Internal.SafeMsQuicConfigurationHandle.Create(QuicOptions options, QUIC_CREDENTIAL_FLAGS flags, X509Certificate certificate, SslStreamCertificateContext certificateContext, List`1 alpnProtocols, CipherSuitesPolicy cipherSuitesPolicy) in C:\git\runtime\src\libraries\System.Net.Quic\src\System\Net\Quic\Implementations\MsQuic\Interop\SafeMsQuicConfigurationHandle.cs:line 237
   at System.Net.Quic.Implementations.MsQuic.Internal.SafeMsQuicConfigurationHandle.Create(QuicClientConnectionOptions options) in 
```

As far as I know there's nothing special about my machine. How should I diagnose this?

### PR DESCRIPTION
**Pull request by rzikm**
Closes #70673.

This PR adds a check for `DisabledByDefault` registry key which would cause the QUIC connection attempts to fail.

### CODE REVIEW

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs [hunk 1]
```diff
@@ -132,9 +132,16 @@ private static bool IsTls13Disabled()
...
                 using var regKey = Registry.LocalMachine.OpenSubKey(key);
 
-                if (regKey is null) return false;
+                if (regKey is null) continue;
 
-                if (regKey.GetValue("Enabled") is int enabled && enabled == 0)
+                if (regKey.GetValue("Enabled") is int enabled)
+                {
+                    if (enabled == 0) return true;
+                    else continue;
```
Comment (stephentoub) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs:140]:
```suggestion
                    continue;
```

**Review by wfurt — APPROVED** [review]
LGTM.  It seems like we have similar logic in PlatformDetection.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs [hunk 2]
```diff
@@ -132,9 +132,16 @@ private static bool IsTls13Disabled()
...
+                }
+
+                // if Enabled key is not present, check the default settings
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs:143]: Do they get returned sorted? Or how do you know that by this point there isn't `Enable` present?
If they cannot be both present going against each other, that's fine, but the comment is slightly misleading.
I don't know how exactly the registry API behaves, so sorry in advance if my question doesn't make sense.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs:143]: > Or how do you know that by this point there isn't Enable present?

We are checking it 4 lines above :D or I don't understand the question.

Unfortunately, it seems that `Enabled` and `DisabledByDefault` are not redundant, some info can be found [at StackOverflow](https://stackoverflow.com/a/51962330). From my understanding of it, the `Enabled` key takes precedence.

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs]: I have taken another look into this, and it seems that _either_ of these can cause the setup to fail, so I will need to improve it further

**Comment by rzikm** [comment]
Since we are touching this, I split client/server detection, so that clients which have TLS1.3 _Server_ disabled can still create outbound connections.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs [hunk 3]
```diff
@@ -56,7 +56,7 @@ public static SafeMsQuicConfigurationHandle Create(QuicClientConnectionOptions o
             {
                 flags |= QUIC_CREDENTIAL_FLAGS.USE_SUPPLIED_CREDENTIALS;
             }
-            return Create(options, flags, certificate: certificate, certificateContext: null, options.ClientAuthenticationOptions?.ApplicationProtocols, options.ClientAuthenticationOptions?.CipherSuitesPolicy);
+            return Create(false, options, flags, certificate: certificate, certificateContext: null, options.ClientAuthenticationOptions?.ApplicationProtocols, options.ClientAuthenticationOptions?.CipherSuitesPolicy);
```
Comment (wfurt) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs:59]: I don't think you need new `bool`. Client will have QUIC_CREDENTIAL_FLAGS.CLIENT set and server not.

**Review by wfurt — APPROVED** [review]
LGTM

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs [hunk 4]
```diff
@@ -120,24 +123,28 @@ static MsQuicApi()
...
+            if (regKey.GetValue("Enabled") is int enabled && enabled == 0)
+            {
+                return true;
+            }
 
-                if (regKey.GetValue("Enabled") is int enabled && enabled == 0)
-                {
-                    return true;
-                }
+            if (regKey.GetValue("DisabledByDefault") is int disabled && disabled == 1)
```
Comment (saul) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs:145]: With this change, `DisabledByDefault=1` with `Enabled=1` is treated as disabled, is that correct? I would expect that `Enabled=1` would overrule `DisabledByDefault`.

Comment (saul) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs:145]: In fact this configuration is explicitly referred to in the documentation:

> In order to override a system default and set a supported (D)TLS or SSL protocol version to the Disabled by default state, create DWORD registry values named "Enabled" and "DisabledByDefault" with a non-zero value under the corresponding version-specific subkey. The following example shows TLS 1.0 server set to the Disabled by default state:
>
> image

https://docs.microsoft.com/en-us/windows-server/security/tls/tls-registry-settings

Comment (rzikm) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs:145]: I tried this combination locally when debugging the original issue and it still led to `SEC_E_ALGORITHM_MISMATCH`

Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Internal/MsQuicApi.cs:145]: That may also depend on protocols we use e.g. `SslProtocols.None` vs `SslProtocols.Tls13`. One defers to OS default and one explicitly asks.

**Pull request merged**